test format_autocomplete & i18n endpoints

risicle opened the pull request:
This sits on top of https://github.com/alphagov/ckan-functional-tests/pull/7 so you can ignore the first couple of commits.

This implements https://trello.com/c/XXCdaC1C and https://trello.com/c/45ZJreNE

risicle committed:
add tests & schema for format_autocomplete endpoint

risicle committed:
schema self-tests: include example response & tests for format_autocomplete

risicle committed:
add test for /api/i18n/ endpoint

risicle committed:
schema self-tests: add example response and test for i18n endpoint

no kidding

kentsanggds reviewed (commented):
As there's not much to the i18n endpoint is it ever called? and if not, there's probably no need for a test for it

risicle commented:
Well, it _is_ called... by something... no idea why.

kentsanggds reviewed (commented):
Still getting some random failures, maybe not directly related to this PR - 

test_search_datasets_by_full_slug_general_term

Fails when it gets this random slug 

random_pkg_slug = '02p-expenditure-over-25k-mar-20'

risicle commented:
:( I'll take a look at that...

BTW the `harvestobject` tests _will_ also occasionally fail because of these mysterious 404s we get for some objects.

risicle commented:
Interestingly, though `02p-expenditure-over-25k-mar-20` isn't findable on staging, it _is_ findable on production. This suggests to me that it is a problem with staging's solr synchronization and something I'd consider a genuine bug.

kentsanggds commented:
There are some issues with the data sync process, which obviously need addressing but there are other works being prioritised.

If some of the tests do occasionally fail then we should build some tolerance to factor this in as we don't want to raise a false alarm that the endpoint is not working when in fact the data is not good.

risicle commented:
Building tolerance into this test is something... well, I'm unsure how I'd do it TBH, I'm not certain it would end up as much of a test, and it's not just _this_ test the index being out of sync could cause to fail. Really, most of the search-related tests are tricky if we're not allowed to make the assumption that the db should be in sync with the index.

I could look at marking different sets of tests which are known to sometimes fail for different reasons so it's possible to disable certain subsets of the test suite...?

kentsanggds commented:
I wonder if it can be something like retrying a test with the random id for 5 times, then we should definitely be able to ensure that it will pass. For the failed tests we can provide a warning but as long as the endpoint eventually passes we can pass it. 

Or identify good ids and test against those only is the other approach, there is a risk that the datasets could change in which case we need to update the tests but this should guarantee the test passing.

risicle commented:
> I wonder if it can be something like retrying a test with the random id for 5 times

That might be an idea... I'll think how i can do that.

risicle committed:
introduce inc_sync_sensitive setting

A variable controlling whether to include tests that are sensitive to
whether the target instance has consistent data across e.g. its database
and search index. We would normally use pytest.mark for this kind of thing,
but that doesn't allow you to mark & skip individual subtests

risicle commented:
Ok, try now with the setting `inc_sync_sensitive` set to `false` in `config.json`.

kentsanggds commented:
Things seem to pass when the value is set, (the setting should probably also be present in `config.json` so it's clear that its an option), but then it also looks like we are skipping some tests, 

I'm worried that the default will be to skip those tests which would then make them obsolete as they would never be run because they are not reliable.

Is there no way to retry a number of times? I was thinking of something like this -

```
pass = False

for i in range(number_of_retries):
  result = test_the_endpoint()
  if result == expected:
    pass = True
    break

assert pass
```

or something like that 😄

risicle commented:
I looked at doing that - I couldn't find a way of doing it in a pytest-y way that doesn't get quite ugly quite quickly - it's only one of the subtests that fails in a way that we want to retry after all, so we'd have to detect that specific failure... or I'd have to turn the test inside-out in a way that would make failures less clear to trace back to their cause. And seeing as I was starting to get flashbacks to retry-loops that I'd written into previous test suites and the dark alleys they lead down, I thought it would be better to let the "fixed data" tests (which I have not yet created) cover this functionality if we want to keep `inc_sync_sensitive` disabled long-term.

kentsanggds reviewed (changes requested):
OK, lets go for the fixed data approach mentioned in this card - https://trello.com/c/mo77VHfw/25-e2e-tests-fixed-data-tests

Also, could the README be updated with `inc_sync_sensitive` config settings and the `inc_sync_sensitive` setting be set in `config.json` with the default value for clarity. Then it's good to go.

risicle commented:
Oh yes, forgot it even had a readme. It still says it runs against production by default.

risicle committed:
README.md: update, add note on inc_sync_sensitive setting

kentsanggds reviewed (approved):
👍
